hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -2325,6 +2332,85 @@ func (r *HostedClusterReconciler) reconcileAutoscaler(ctx context.Context, creat
 	return autoscaler.ReconcileAutoscaler(ctx, r.Client, hcp, clusterAutoscalerImage, utilitiesImage, createOrUpdate, r.SetDefaultSecurityContext, config.MutatingOwnerRefFromHCP(hcp, releaseVersion))
 }
 
+func (r *HostedClusterReconciler) reconcileETCDMemberRecovery(ctx context.Context, hcluster *hyperv1.HostedCluster) error {
+	// Reconcile the ETCD member recovery job
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("reconciling etcd failling member recovery")
+	hcpNS := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name)
+
+	podList := &corev1.PodList{}
+	if err := r.List(ctx, podList, client.InNamespace(hcpNS), client.MatchingLabels{
+		"app": "etcd",
+	}); err != nil {
+		return fmt.Errorf("failed to list etcd pods: %w", err)
+	}
+
+	var failingPods []corev1.Pod
+	for _, pod := range podList.Items {
+		for _, containerStatus := range pod.Status.ContainerStatuses {
+			if containerStatus.State.Waiting != nil && containerStatus.State.Waiting.Reason == "CrashLoopBackOff" {
+				failingPods = append(failingPods, pod)
+			}
+		}
+	}
+
+	if len(failingPods) > 1 {
+		return fmt.Errorf("more than one etcd pod is failing, skipping recovery procedure because lack of quorum: %v", failingPods)
+	}
+
+	if len(failingPods) == 0 {
+		log.Info("no etcd pod is failing, skipping recovery procedure and deleting annotation", "annotation", ETCDRestoreFailingMemberAnnotation)
+		hclusterCP := hcluster.DeepCopy()
+		if _, exists := hclusterCP.Annotations[ETCDRestoreFailingMemberAnnotation]; exists {
+			delete(hclusterCP.Annotations, ETCDRestoreFailingMemberAnnotation)
+			if err := r.Update(ctx, hclusterCP); err != nil {
+				return fmt.Errorf("failed to update hosted cluster: %w", err)
+			}
+		}
+		return nil
+	}
+
+	etcdPV := &corev1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("data-%s", failingPods[0].Name),
+			Namespace: hcpNS,
+		},
+	}
+
+	if err := r.Get(ctx, client.ObjectKeyFromObject(etcdPV), etcdPV); err != nil {
+		return fmt.Errorf("failed to get etcd persistent volume: %w", err)
+	}
+
+	if etcdPV.Spec.ClaimRef == nil {
+		return fmt.Errorf("etcd persistent volume has no claim reference: %v", etcdPV)
+	}
+
+	etcdPVC := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      etcdPV.Spec.ClaimRef.Name,
+			Namespace: hcpNS,
+		},
+	}
+
+	if err := r.Get(ctx, client.ObjectKeyFromObject(etcdPVC), etcdPVC); err != nil {
+		return fmt.Errorf("failed to get etcd persistent volume claim: %w", err)
```

Comment:
I'm trying to think of there are cases when we fail to get the PVC or the PV, yet we would still like to proceed and attempt the pod deletion?